rust/src/tor.rs

```diff
@@ -0,0 +1,255 @@
+//! Tor support
+
+use std::{fmt, future::Future, io, path::Path, sync::Arc};
+
+use arti_client::{
+    config::{ConfigBuildError, TorClientConfigBuilder},
+    TorClient,
+};
+use http_body_util::{BodyExt, Empty};
+use hyper::{
+    body::{Buf, Bytes, Incoming},
+    client::conn,
+    http::{request::Builder, uri::Scheme},
+    Request, Response, Uri,
+};
+use hyper_util::rt::TokioIo;
+use serde::de::DeserializeOwned;
+use tokio::io::{AsyncRead, AsyncWrite};
+use tokio_rustls::{
+    rustls::{ClientConfig, OwnedTrustAnchor, RootCertStore, ServerName},
+    TlsConnector,
+};
+use tor_rtcompat::{BlockOn, PreferredRuntime};
+use tracing::{debug, error};
+
+mod exchange_rate;
+
+pub struct TorRuntime {
+    runtime: PreferredRuntime,
+    client: Client,
+}
+
+impl TorRuntime {
+    #[tracing::instrument]
+    pub(crate) fn create(tor_dir: &Path) -> Result<Self, Error> {
+        let runtime = PreferredRuntime::create()?;
+        let client = runtime.block_on(async { Client::create(tor_dir).await })?;
+        Ok(Self { runtime, client })
+    }
+
+    pub(crate) fn runtime(&self) -> &PreferredRuntime {
+        &self.runtime
+    }
+
+    pub(crate) fn client(&self) -> &Client {
+        &self.client
+    }
+}
+
+pub(crate) struct Client {
+    inner: TorClient<PreferredRuntime>,
+}
+
+impl Client {
+    #[tracing::instrument]
+    pub(crate) async fn create(tor_dir: &Path) -> Result<Self, Error> {
+        let runtime = PreferredRuntime::current()?;
+
+        if !tokio::fs::try_exists(tor_dir).await? {
+            return Err(Error::MissingTorDirectory);
+        }
+
+        let mut config_builder = TorClientConfigBuilder::from_directories(
+            tor_dir.join("arti-data"),
+            tor_dir.join("arti-cache"),
+        );
+        config_builder
+            .storage()
+            .permissions()
+            .dangerously_trust_everyone();
```

Comment:
The permissions here need testing. I was having trouble getting Arti to work in the iOS simulator, and one of the things I did to get around it was enable this setting. I eventually also figured out that on the Swift side we needed to pass over `URL.path` instead of `URL.absoluteString`, but I haven't gone back and experimented with other permissions here yet.

Reply:
This was removed after we moved this code to `zcash_client_backend`, after I confirmed it wasn't necessary.